### PR TITLE
EVA 618 Parametrize the use of mapping files for VCF

### PR DESCRIPTION
**Pull request by jorizci**
Added new parameter to load a mapping file to interpret the info of VCF. 

Is there any VCF that can be used as a test for this characteristic?

### CODE REVIEW

File: src/main/java/uk/ac/ebi/eva/pipeline/io/mappers/AggregatedVcfLineMapper.java [hunk 1]
```diff
@@ -34,18 +39,26 @@
...
+                                   Optional<String> mappingFilePath) {
+        Assert.notNull(fileId);
+        Assert.notNull(studyId);
+        Assert.notNull(aggregation);
+
         this.fileId = fileId;
         this.studyId = studyId;
+
+        Properties mappings = mappingFilePath.map(path -> FileUtils.getPropertiesFile(path).orElseThrow
+                (RuntimeException::new)).orElse(null);
```
Comment (jmmut) [inline src/main/java/uk/ac/ebi/eva/pipeline/io/mappers/AggregatedVcfLineMapper.java]: funny line break here. can we make it more readable?

File: src/main/java/uk/ac/ebi/eva/utils/FileUtils.java [hunk 2]
```diff
@@ -1,5 +1,7 @@
 package uk.ac.ebi.eva.utils;
```
Comment (jmmut) [inline src/main/java/uk/ac/ebi/eva/utils/FileUtils.java]: license

File: src/main/java/uk/ac/ebi/eva/utils/FileUtils.java [hunk 3]
```diff
@@ -41,4 +47,16 @@ public static Resource getResource(File file) throws IOException {
...
         return resource;
     }
+
+    public static Optional<Properties> getPropertiesFile(String propertiesFilePath){
+        try {
+            InputStream input = new FileInputStream(propertiesFilePath);
+            Properties properties = new Properties();
+            properties.load(input);
+            return Optional.ofNullable(properties);
+        }catch(Exception e){
```
Comment (jorizci) [inline src/main/java/uk/ac/ebi/eva/utils/FileUtils.java]: Spaaaaceeeeeeeeee!

File: src/main/java/uk/ac/ebi/eva/utils/FileUtils.java (hunk 2 above)
Comment (jorizci) [inline src/main/java/uk/ac/ebi/eva/utils/FileUtils.java]: Cof cof this file has changes from Mut cof cof... and he didn't put the license cof cof...

File: src/main/java/uk/ac/ebi/eva/pipeline/parameters/JobParametersNames.java [hunk 4]
```diff
@@ -136,4 +136,6 @@
     public static final String CONFIG_RESTARTABILITY_ALLOW = "config.restartability.allow";
 
     public static final String CONFIG_CHUNK_SIZE = "config.chunk.size";
+
+    public static final String AGGREGATED_MAPPING_FILE = "aggregated.mapping.file";
```
Comment (cyenyxe) [inline src/main/java/uk/ac/ebi/eva/pipeline/parameters/JobParametersNames.java]: Wouldn't it be more intuitive to name this in a similar fashion to the aggregation mode?

[This document](https://docs.google.com/document/d/1ZMhLC-Mmd_1ygaFsdIRbcubJ9t4goEFnaLswk71v2Nw/) suggests `input.file.aggregation.mapping`, but given how the rest of arguments look then `input.vcf.aggregation.mapping-path` could be better.

File: src/main/java/uk/ac/ebi/eva/utils/ThrowingFunction.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+import java.util.function.Function;
+
+/**
+ * Functional interface to use functions that throw exceptions inside functional code.
+ *
+ * @param <T> Accepted type
+ * @param <R> Return type
+ */
+@FunctionalInterface
+public interface ThrowingFunction<T, R> extends Function<T, R> {
```
Comment (cyenyxe) [inline src/main/java/uk/ac/ebi/eva/utils/ThrowingFunction.java]: If the use of Optional is removed then this class is no longer necessary.

File: src/main/java/uk/ac/ebi/eva/pipeline/configuration/readers/VcfReaderConfiguration.java [hunk 6]
```diff
@@ -54,11 +55,13 @@
...
         String studyId = parameters.getStudyId();
+        File vcfFile = new File(parameters.getVcf());
 
         if (VariantSource.Aggregation.NONE.equals(parameters.getVcfAggregation())) {
-            return new VcfReader(fileId, studyId, parameters.getVcf());
+            return new VcfReader(fileId, studyId, vcfFile);
         } else {
-            return new AggregatedVcfReader(fileId, studyId, parameters.getVcfAggregation(), parameters.getVcf());
+            return new AggregatedVcfReader(fileId, studyId, parameters.getVcfAggregation(),
+                    parameters.getAggregatedMappingFile(), vcfFile);
```
Comment (cyenyxe) [inline src/main/java/uk/ac/ebi/eva/pipeline/configuration/readers/VcfReaderConfiguration.java]: This should read from default mappings in those cases where the mapping file is not specified. See the [`src/main/resources` folder from Biodata](https://github.com/EBIvariation/biodata/tree/hotfix/0.4/biodata-models/src/main/resources) for the 2 necessary mapping files, one for EXAC and one for EVS mode.

File: src/test/java/uk/ac/ebi/eva/pipeline/parameters/validation/InputVcfAggregationMappingPathTest.java [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    @Rule
+    public PipelineTemporaryFolderRule temporaryFolder = new PipelineTemporaryFolderRule();
+
+    @Before
+    public void setUp() throws Exception {
+        validator = new InputVcfAggregationMappingPath();
+    }
+
+    @Test
+    public void inputFastaIsValid() throws JobParametersInvalidException, IOException {
```
Comment (jmmut) [inline src/test/java/uk/ac/ebi/eva/pipeline/parameters/validation/InputVcfAggregationMappingPathTest.java]: input fasta? applies to all the methods here

File: src/main/java/uk/ac/ebi/eva/pipeline/parameters/validation/InputVcfAggregationMappingPath.java [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+import org.springframework.batch.core.JobParametersInvalidException;
+import org.springframework.batch.core.JobParametersValidator;
+import uk.ac.ebi.eva.pipeline.parameters.JobParametersNames;
+
+/**
+ * Checks that the aggregated mapping file exists and is readable.
+ *
+ * @throws JobParametersInvalidException If the file is not a valid path, does not exist or is not readable
+ */
+public class InputVcfAggregationMappingPath implements JobParametersValidator {
```
Comment (jmmut) [inline src/main/java/uk/ac/ebi/eva/pipeline/parameters/validation/InputVcfAggregationMappingPath.java]: This should be called InputVcfAggregationMappingPathValidator

File: src/main/java/uk/ac/ebi/eva/pipeline/io/mappers/VariantAggregatedVcfFactory.java [hunk 9]
```diff
@@ -83,8 +83,20 @@ public VariantAggregatedVcfFactory() {
      *               be a bijection, i.e. there must not be repeated entries in any side. The part before the '.' can be any string
      *               naming the group. The part after the '.' must be one of AF, AC, AN or GTC.
      */
-    public VariantAggregatedVcfFactory(Properties tagMap) {
-        this.tagMap = tagMap;
+    public VariantAggregatedVcfFactory(Properties mappings) {
+        if(mappings == null){
+            loadDefaultMappings();
+        }else{
```
Comment (jmmut) [inline src/main/java/uk/ac/ebi/eva/pipeline/io/mappers/VariantAggregatedVcfFactory.java]: what do you have against spaces? :P

**Pull request merged**